cloudmarker/clouds/azkv.py

```diff
@@ -0,0 +1,304 @@
+
+                # Break after pulling data for self._max_recs number
+                # of Key Vault for a subscriber. Note that if
+                # self._max_recs is 0 or less, then the following
+                # condition never evaluates to True.
+                if key_vault_index + 1 == self._max_recs:
+                    _log.info('Stopping Key Vault fetch due '
+                              'to _max_recs: %d; %s', self._max_recs,
+                              util.outline_az_sub(sub_index, sub,
+                                                  self._tenant))
+                    break
+        except CloudError as e:
+            _log.error('Failed to fetch Key Vault; %s; error: %s: %s',
+                       util.outline_az_sub(sub_index, sub, tenant),
+                       type(e).__name__, e)
+
+    def _process_key_vault(self, key_vault_index, key_vault_name, rg_name,
+                           sub_index, sub):
+        """Get details of Key Vault in management and data plane.
+
+        Arguments:
+            sub_index (int): Subscription index (for logging only).
+            sub (Subscription): Azure subscription object.
+            rg_name (str): Resource group name.
+            key_vault_index (int): Key Vault index (for logging only).
+            key_vault_name (str): Name of the Key Vault.
+
+        Yields:
+            dict: An Azure Key Vault server record.
+
+        """
+        def _auth_callback(server, resource, scope):
+            credentials = self._key_vault_credentials
+            token = credentials.token
+            return token['token_type'], token['access_token']
+
+        _log.info('Working on Key Vault #%d: %s; %s', key_vault_index,
+                  key_vault_name, util.outline_az_sub(sub_index, sub,
+                                                      self._tenant))
+        subscription_id = sub.get('subscription_id')
+        creds = self._credentials
+        key_vault_mgmt_client = KeyVaultManagementClient(creds,
+                                                         subscription_id)
+        key_vault_details = key_vault_mgmt_client.vaults.get(rg_name,
+                                                             key_vault_name)
+        yield from _get_normalized_key_vault_record(key_vault_details, sub)
+        try:
+            kv_client = \
+                KeyVaultClient(KeyVaultAuthentication(_auth_callback))
+            secrets = \
+                kv_client.get_secrets(key_vault_details.properties.vault_uri)
+            keys = \
+                kv_client.get_keys(key_vault_details.properties.vault_uri)
+            # Retrieve data using each iterator.
+            for record in itertools.chain(
+                    _get_data_record(secrets, 'key_vault_secret',
+                                     sub),
+                    _get_data_record(keys, 'key_vault_key',
+                                     sub),
+            ):
+                yield record
```

Comment:
Right now, this part of the code logs many `KeyVaultErrorException` errors when there are access issues. I suggest making these logs less noisy and handling this exceptions explicitly like this:

```python
            try:
                for record in _get_data_record(secrets, 'key_vault_secret', sub):
                    yield record
            except ClientException as e:
                _log.error('Failed to fetch key details for key_vault; '
                           '%s; error: %s: %s',
                           util.outline_az_sub(sub_index, sub,
                           self._tenant), type(e).__name__, e)

            try:
                for record in _get_data_record(keys, 'key_vault_key', sub):
                    yield record
            except ClientException as e:
                _log.error('Failed to fetch key details for key_vault; '
                           '%s; error: %s: %s',
                           util.outline_az_sub(sub_index, sub,
                           self._tenant), type(e).__name__, e)
```

I have chosen `ClientException` in the above example because it is an ancestor class of `KeyVaultErrorException` and is easily importable like this:

```python
from msrest.exceptions import ClientException
```

I see two advantages of this:

- Logs would be less noisy like I explained before. The entire stack traces won't be logged. Only the exception, exception message, and our log message.
- Even if `_get_data_record(..., 'key_vault_secret', ...)` fails, we will give a chance for `_get_data_record(..., 'key_vault_key', ...)` to execute.

Also, note that it would be nice to use the `util.outline_az_sub()` method because it formats contextual information like subscription and tenant information neatly.

Reply:
Fixed